File: HashmapAndSet/group-anagrams.js
/*Given an array of strings strs, group the anagrams together.

For example, given strs = ["eat","tea","tan","ate","nat","bat"], return [["bat"],["nat","tan"],["ate","eat","tea"]].*/

const groupAnagrams = (strs) => {
    const groups = new Map();

    for (let str of strs) {
        const key = str.split('').sort().join('');

        if (!groups.has(key)) {
            groups.set(key, []);
        }
            
        groups.get(key).push(str);
    }

    const ans = [];

    for (let group of groups.values()) {
        ans.push(group);
    }

    return ans;
}

/*Given n as the length of strs and m as the average length of the strings, we iterate over each string and sort it, which costs O(n⋅m⋅log⁡m). 
Then, we need to iterate over the keys. I
n the worst case scenario, when there are no matching anagrams, there will be n groups, which means this will cost O(n), 
giving an overall time complexity of O(n⋅m⋅log⁡m) (the final  +n is dominated). 

The space complexity is O(n⋅m) as each string will be placed in an array within the hash map.*/

const strs = ["eat","tea","tan","ate","nat","bat"];
console.log(groupAnagrams(strs));

File: HashmapAndSet/group_anagrams.py
from collections import defaultdict
from typing import List


class GroupAnagrams:
    def groupAnagramsTuple(self, strs: List[str]) -> List[List[str]]:
        groups = defaultdict(list)
        for s in strs:
            freq = [0] * 26
            for c in s:
                freq[ord(c) - ord("a")] += 1
            freq_tuple = tuple(freq)
            groups[freq_tuple].append(s)

        return list(groups.values())

    # Time complexity: O(n * m) where n is the number of strings in the input array 
    # and m is the maximum length of a string in the input array.
    # We iterate through the input array once, and for each string, 
    # we compute its frequency tuple in O(m) time.
    # Space complexity: O(n * m) in the worst case where all strings are unique and have a length of m, 
    # resulting in n unique frequency tuples stored in the hashmap.

    def groupAnagramsSorted(self, strs: List[str]) -> List[List[str]]:
        groups = defaultdict(list)
        for s in strs:
            s_sorted = "".join(sorted(list(s)))
            groups[s_sorted].append(s)

        return list(groups.values())

    # Time complexity: O(n * m log m) where n is the number of strings in the input array
    # and m is the maximum length of a string in the input array.
    # We iterate through the input array once, and for each string, we sort it in
    # O(m log m) time.
    # Space complexity: O(n * m) in the worst case where all strings are unique
    # and have a length of m, resulting in n unique sorted strings stored in the hashmap.